protocols/bftcosi/bftcosi.go

```diff
@@ -0,0 +1,615 @@
+	return err
+}
+
+// startPrepareChallenge create the challenge and send its down the tree
+func (bft *ProtocolBFTCoSi) startChallengePrepare() error {
+	// create challenge from message's hash:
+	prep := &msgPrepare{Msg: bft.Msg}
+	h := prep.Hash()
+	ch, err := bft.prepare.CreateChallenge(h)
+	if err != nil {
+		return err
+	}
+
+	bftChal := &ChallengePrepare{
+		TYPE:      RoundPrepare,
+		Challenge: ch,
+		Msg:       bft.Msg,
+	}
+
+	go bft.verificationFun(bft.Msg, bft.verifyChan)
+
+	dbg.Lvl3(bft.Name(), "BFTCoSi Start Challenge PREPARE")
+	// send to children
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, bftChal)
+	}
+	return err
+}
+
+// startCommitChallenge waits the end of the "prepare" round.
+// Then it creates the challenge and sends it along with the
+// "prepare" signature down the tree.
+func (bft *ProtocolBFTCoSi) startChallengeCommit() error {
+	c, err := bft.commit.CreateChallenge(bft.Msg)
+	if err != nil {
+		return err
+	}
+
+	// send challenge + signature
+	cc := &ChallengeCommit{
+		TYPE:      RoundCommit,
+		Challenge: c,
+		Signature: bft.prepare.Signature(),
+	}
+	dbg.Lvl3("BFTCoSi Start Challenge COMMIT")
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, cc)
+	}
+	return err
+}
+
+// handlePrepareChallenge receive the challenge messages for the "prepare"
+// round.
+func (bft *ProtocolBFTCoSi) handleChallengePrepare(ch *ChallengePrepare) error {
+	bft.Msg = ch.Msg
+	// start the verification of the message
+	go bft.verificationFun(bft.Msg, bft.verifyChan)
+	// acknowledge the challenge and send its down
+	chal := bft.prepare.Challenge(ch.Challenge)
+	ch.Challenge = chal
+
+	dbg.Lvl3(bft.Name(), "BFTCoSi handle Challenge PREPARE")
+	// go to response if leaf
+	if bft.IsLeaf() {
+		return bft.startResponsePrepare()
+	}
+	var err error
+	for _, tn := range bft.Children() {
+		err = bft.SendTo(tn, ch)
+	}
+	return err
+}
+
+// handleCommitChallenge will verify the signature + check if no more than 1/3
+// of participants refused to sign.
+func (bft *ProtocolBFTCoSi) handleChallengeCommit(ch *ChallengeCommit) error {
+	ch.Challenge = bft.commit.Challenge(ch.Challenge)
+	prep := &msgPrepare{Msg: bft.Msg}
+	h := prep.Hash()
+
+	// verify if the signature is correct
+	if err := cosi.VerifyCosiSignatureWithException(bft.suite,
+		bft.aggregatedPublic, h, ch.Signature,
+		ch.Exceptions); err != nil {
+		dbg.Error(bft.Name(), "Verification of the signature failed:", err)
+		bft.signRefusal = true
+	}
+
+	// Check if we have no more than 1/3 failed nodes
+	if len(ch.Exceptions) > int(bft.threshold) {
+		dbg.Errorf("More than 1/3 (%d/%d) refused to sign ! ABORT",
+			len(ch.Exceptions), len(bft.EntityList().List))
+		bft.signRefusal = true
+	}
+
+	// store the exceptions for later usage
+	bft.tempExceptions = ch.Exceptions
+	dbg.Lvl3("BFTCoSi handle Challenge COMMIT")
+	if bft.IsLeaf() {
+		return bft.startResponseCommit()
+	}
+
+	// send it down
+	for _, tn := range bft.Children() {
```

Comment:
Hmm, on my local test this doesn't change anything. But in reality (and maybe in other places besides the one pointed out), sometimes you would want to continue sending even if one send failed (on child went offline). I still changed it like you proposed (almost) everywhere. But I'm sure we will need another function, which continues sending soon.